web_src/js/features/gitgraph.js

```diff
@@ -1,13 +1,579 @@
+  setLeftFlow(column, currentRow, previousRow) {
+    // This is: '----------.' or the like
+    //          '   \  \  /|\'
+
+    // Find the end of the '-' or nearest '/|\' in the previousRow :
+    let originalColumn = column;
+    let flow;
+    for (; column < currentRow.length && currentRow[column] === '-'; column++) {
+      if (column > 0 && column - 1 < previousRow.length && previousRow[column - 1] === '/') {
+        flow = this.takePreviousFlow(column - 1);
+        break;
+      } else if (column < previousRow.length && previousRow[column] === '|') {
+        flow = this.takePreviousFlow(column);
+        break;
+      } else if (
+        column + 1 < previousRow.length &&
+        previousRow[column + 1] === '\\'
+      ) {
+        flow = this.takePreviousFlow(column + 1);
+        break;
+      }
+    }
+
+    // if we have a flow then we found a '/|\' in the previousRow
+    if (flow) {
+      for (; originalColumn < column + 1; originalColumn++) {
+        this.currentFlows[originalColumn] = flow;
+      }
+      return column;
+    }
+
+    // If the symbol in the column is not a '.' then there's likely an error
+    if (currentRow[column] !== '.') {
+      // It really should end in a '.' but this one doesn't...
+      // 1. Step back - we don't want to eat this column
+      column--;
+      // 2. Generate a new flow and await bug-reports...
+      this.currentFlows[column] = this.generateNewFlow(column);
+
+      // 3. Assign all of the '-' to the same flow.
+      for (; originalColumn < column; originalColumn++) {
+        this.currentFlows[originalColumn] = this.currentFlows[column];
+      }
+      return column;
+    }
+
+    // We have a terminal '.' eg. the current row looks like '----.'
+    // the previous row should look like one of '/|\' eg.    '     \'
+    if (column > 0 && column - 1 < previousRow.length && previousRow[column - 1] === '/') {
+      flow = this.takePreviousFlow(column - 1);
+    } else if (column < previousRow.length && previousRow[column] === '|') {
+      flow = this.takePreviousFlow(column);
+    } else if (
+      column + 1 < previousRow.length &&
+      previousRow[column + 1] === '\\'
+    ) {
+      flow = this.takePreviousFlow(column + 1);
+    } else {
+      // Again unexpected so let's generate and wait the bug-report
+      flow = this.generateNewFlow(column);
+    }
+
+    // Assign all of the rest of the ----. to this flow.
+    for (; originalColumn < column + 1; originalColumn++) {
+      this.currentFlows[originalColumn] = flow;
+    }
+
+    return column;
+  }
+}
+
+function generateRandomColorString() {
+  const chars = '0123456789ABCDEF';
+  const stringLength = 6;
+  let randomString = '',
+    rnum,
+    i;
+  for (i = 0; i < stringLength; i++) {
+    rnum = Math.floor(Math.random() * chars.length);
+    randomString += chars.substring(rnum, rnum + 1);
+  }
+
+  return randomString;
+}
+
 export default async function initGitGraph() {
```

Comment:
Would prefer this to stay at the top of the file.

Reply:
Can't as eslint then moans that you're using the function before it's defined...